README
This program takes a .drg file from I-Doser and convert it to an .sbg SBaGen file.

The authors of this project have nothing to do with I-Doser, I-Doser is a registered
trademark.

This project does not supports piracy or any illegal action, the use of this product 
must be stricly to pass .drg files to .sbg for later convertion to .wav or similar 
portable format, for those who (as the author) doesn't have a Windows pc to play the
dose.

This project will never be possible without the help of "The Unofficial I-Doser Blog"
and their _steAlth_'s DRG Author by konstAnt and _khAttAm_ to tell the truth this is
a rip off stripped version of konstAnt and _KhAttAm_ _steAlth_'s DRG Author. They did
all the work with the .drg files, I just port some parts of it to C.

Related links:
* http://uazu.net/sbagen/ (Main Engine, the binaural generator)
* http://www.i-doser.com/ (The company who makes the .drg files)
* http://www.i-doser.com/store/ (The place to buy good .drg files)
* http://theidoserblog.blogspot.com/ (The guys who made this possible)
* http://p4c0.wordpress.com/ (Useless blog)


-- 
Manuel Argüelles


ChangeLog
2012-07-03  Manuel Argüelles  <manuel.arguelles@gmail.com>

	This file will not be used from now on, if  you want to list the
	changelog execute: git log

2012-06-29  Manuel Argüelles  <manuel.arguelles@gmail.com>

	Fix reported issue #3: 80 columns base64 data is not supported by
	openssl, indentation fixed (tabs removed), copyright heades added.

	* src/drgtosbg.c
	* src/drgdata.c
	* src/drgdata.h

2009-06-07  Manuel Argüelles  <manuel.arguelles@azetti.com>

	* src/Makefile.am (drg2sbg_SOURCES): Remove -Wall -Werror flags.

	* src/drgtosbg.c (print_formated): Removed -i flag, now information
	is printed in the output.

2009-06-06  Manuel Argüelles  <manuel.arguelles@azetti.com>

	* configure.ac: Replaced old configure.in

2008-07-20	Manuel Arguelles        <manuel.arguelles@gmail.com>
	* Changed DRG information output to fit 50 characters instead of
	  everything in one line.

2008-06-21	Manuel Arguelles	<manuel.arguelles@gmail.com>
	* Initial sketch
	
	* Added information to README and AUTHORS files. TODO file added

	* Changed name of executable from drgtosbg to drg2sbg to match
	  project name

	* Added version control


src/drgdata.h
/* 
 * Copyright (C) 2012  Manuel Argüelles <manuel.arguelles@gmail.com>
 *
 * This file is part of drg2sbg.
 *
 * Drg2sbg is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 *
 * Foobar is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Foobar.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

#ifndef DRG_DATA_H
#define DRG_DATA_H


enum drg_elements {
    HEADER = 0,
    TITLE,
    IMAGE,
    INFO,
    SBG_DATA,
    MAX_ELEMENTS
};

typedef struct drgdata_ DrgData; 


DrgData *drg_data_new(void);

void drg_data_free(DrgData *drg);

void drg_add_byte(DrgData *drg, int element, int byte);

char *drg_get_uncoded_data(DrgData *drg, int element);

#endif /* DRG_DATA_H */


src/drgtosbg.c
/* 
 * Copyright (C) 2012  Manuel Argüelles <manuel.arguelles@gmail.com>
 *
 * This file is part of drg2sbg.
 *
 * Drg2sbg is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 *
 * Foobar is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Foobar.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <errno.h>
#include <string.h>
#include <locale.h>

#include "drgdata.h"
#include "config.h"

/* Prototypes */
static void print_usage(char *prog_name);
static void print_raw_usage(char *prog_name);
static int output_file_idx(int argc, char *argv[]);
static int input_file_idx(int argc, char *argv[]);
static int print_version(int argc, char *argv[]);
static int raw_output_idx(int argc, char *argv[]);
static void print_formated(FILE *out, const char *string, int line_len);
static void print_raw(FILE *out, DrgData *drg, int element);


static void print_usage(char *prog_name)
{
    fprintf(stderr, "please use: %s [options] drgfile\n", prog_name);
    fprintf(stderr, "where options are:\n");
    fprintf(stderr, "   -v         Print program version and exit\n");
    fprintf(stderr, "   -o file    Write to file (default to stdout)\n");
    fprintf(stderr, "   -r element Output raw element\n");
    fprintf(stderr, "\n");
}

static void print_raw_usage(char *prog_name)
{
    fprintf(stderr, "when using %s with '-r element' option\n", prog_name);
    fprintf(stderr, "element must be one of the drgfile elements:\n");
    fprintf(stderr, "   1  Header\n");
    fprintf(stderr, "   2  Title\n");
    fprintf(stderr, "   3  Image\n");
    fprintf(stderr, "   4  Description\n");
    fprintf(stderr, "   5  Sbagen data\n");
    fprintf(stderr, "\n");
}

static int output_file_idx(int argc, char *argv[])
{
    int i = 1;
    while (i < argc) {
        if (strcmp(argv[i], "-o") == 0 || strcmp(argv[i], "--output") == 0)
            return ++i;
        i++;
    }
    return -1;
}

static int input_file_idx(int argc, char *argv[])
{
    int i = 1;
    while (i < argc) {
        if (strcmp(argv[i], "-o") == 0 || strcmp(argv[i], "--output") == 0)
            i++;
        else if (strcmp(argv[i], "-r") == 0 || strcmp(argv[i], "--raw") == 0)
            i++;
        else if (strcmp(argv[i], "-v") == 0 || strcmp(argv[i], "--version") == 0)
            i = i; 
        else 
            return i;
        i++;
    }
    return -1;
}

static int raw_output_idx(int argc, char *argv[])
{
    int i = 1;
    while (i < argc) {
        if (strcmp(argv[i], "-r") == 0 || strcmp(argv[i], "--raw") == 0)
            return ++i;
        i++;
    }
    return -1;
        
}

static int print_version(int argc, char *argv[])
{
    int i = 1;
    while (i < argc) {
        if (strcmp(argv[i], "-v")==0 || strcmp(argv[i], "--version")==0)
            return 1;
        i++;
    }
    return 0;   
}

static void print_formated(FILE *out, const char *string, int line_len)
{
    size_t size;
    size_t i;
    size_t cline = 1;

    if (string == NULL) 
        return;

    size = strlen(string);
    if (size == 0)
        return;
    
    fprintf(out, "## ");
    for (i = 0; i < size; i++) {
        if (string[i] == '\n') { 
            cline = 1;
            fprintf(out, "\n## ");
            continue;
        } else if (string[i] == ' ' && cline >= line_len) {
            cline = 1;
            fprintf(out, "\n## ");
            continue;
        }
        fputc(string[i], out);
        cline++;
    }
    fputc('\n', out);

}

static void print_raw(FILE *out, DrgData *drg, int element)
{
    char *output = NULL;

    output = drg_get_uncoded_data(drg, element);

    if (output && strlen(output) > 0)
        fprintf(out, "%s\n", output);
    
    free(output);
}

int main(int argc, char *argv[])
{
    FILE *drg_fp;
    FILE *sbg_fp = NULL;
    DrgData *drg;
    int c, i = 0; 
    int raw = 0;
    char *output;

    setlocale(LC_ALL, "");

    i = print_version(argc, argv);
    if (i) {
        fprintf(stdout, "\n%s VERSION: %s\n\n", PACKAGE, VERSION);
        return 0;
    }
        
    i = input_file_idx(argc, argv);
    if (i > 0) {
        drg_fp = fopen(argv[i], "r");
        if (drg_fp == NULL) {
            fprintf(stderr, "could not open file %s: %s\n", argv[i], strerror(errno));
            return 1;
        }
    } else {
        print_usage(argv[0]);
        return 1;
    }

    i = output_file_idx(argc, argv);
    if (i > 0) {
        if (i < argc) {
            sbg_fp = fopen(argv[i], "w");
            if (sbg_fp == NULL) {
                fprintf(stderr, "could not open output file %s: %s\n", argv[i], strerror(errno));
                return 1;
            }
        } else {
            fprintf(stderr, "could not open output file (not specified)\n");
            return 1;
        }
    }

    i = raw_output_idx(argc, argv);
    if (i > 0) {
        if (i < argc) {
            raw = atoi(argv[i]);
            if (raw < 1 || raw > 5) {
                print_raw_usage(argv[0]);
                return 1;
            }
        } else {
            print_raw_usage(argv[0]);
            return 1;
        }        
    }
    i = 0;

    drg = drg_data_new();
    /* The header is a special element, not separated by @ */
    while (( c = fgetc(drg_fp)) != EOF) {
        if (c == '\n' || c == '\r' || c == '@')
            break;
        else
            drg_add_byte(drg, i, c);
    }

    i = 1;
    /* Rest of elements separated by @ */
    while ((c = fgetc(drg_fp)) != EOF) {
        if (c == '@') {
            i++;
        } else if (c != '\n' && c != '\r') {
            drg_add_byte(drg, i, c);
        }
    }
    fclose(drg_fp);

    
    if (sbg_fp == NULL)
        sbg_fp = stdout;

    if (raw == 0) {
        output = drg_get_uncoded_data(drg, INFO);
        print_formated(sbg_fp, output, 50);
        free(output);

        output = drg_get_uncoded_data(drg, SBG_DATA);
        if (output == NULL) {
            fprintf(stderr, "Error decoding drg file\n");   
            drg_data_free(drg);
            return 1;
        }

        fprintf(sbg_fp, "\n-SE\n%s\n", output);
        free(output);   
    } else {
        print_raw(sbg_fp, drg, raw - 1);
    }

    if (sbg_fp != stdout) 
        fclose(sbg_fp);
    
    drg_data_free(drg);
    
    return 0;
}




src/drgdata.c
/* 
 * Copyright (C) 2012  Manuel Argüelles <manuel.arguelles@gmail.com>
 *
 * This file is part of drg2sbg.
 *
 * Drg2sbg is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 *
 * Foobar is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Foobar.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <errno.h>
#include <string.h>
#include <assert.h>
#include "drgdata.h"

#include <openssl/bio.h>
#include <openssl/evp.h>
#include <openssl/buffer.h>

struct drgdata_ {
    unsigned char *data[MAX_ELEMENTS];
    int len[MAX_ELEMENTS];
    int alloc[MAX_ELEMENTS];
};

/* Prototypes */
static const char *element_to_text(int element);


static const char *element_to_text(int element)
{
    char *str;

    switch (element) {
    case HEADER:
        str = "header section";
        break;
    case TITLE:
        str = "title section";
        break;
    case IMAGE:
        str = "image section";
        break;
    case INFO:
        str = "information section";
        break;
    case SBG_DATA:
        str = "sbagen section";
        break;
    default:
        str = "unknown section";
        break;
    }

    return str;
}

DrgData *drg_data_new(void) 
{
    int i;
    DrgData *drg;
    drg = malloc(sizeof(DrgData));
    if (drg == NULL) {
        return NULL;
    }

    for (i = 0; i < MAX_ELEMENTS; i++) {
        drg->data[i] = malloc(1024);
        drg->len[i] = 0;
        drg->alloc[i] = 1024;
    }
    
    return drg;
}

void drg_data_free(DrgData *drg)
{
    int i;
    for (i = 0; i < MAX_ELEMENTS; i++) {
        free(drg->data[i]);
    }
    free(drg);
}

void drg_add_byte(DrgData *drg, int element, int byte)
{
    assert(drg != NULL);
    if (element >= MAX_ELEMENTS)
        return;

    if (drg->len[element] < drg->alloc[element]) {
        drg->data[element][drg->len[element]] = (unsigned char)byte;
        drg->len[element]++;
    } else {
        unsigned char *new;
        drg->alloc[element]+=1024;
        new = realloc(drg->data[element], drg->alloc[element]);
        drg->data[element] = new;
        drg->data[element][drg->len[element]] = (unsigned char)byte;
        drg->len[element]++;
    }
}

char *drg_get_uncoded_data(DrgData *drg, int element)
{
    BIO *b64;
    BIO *mem;
    char *data;
    unsigned char temp;
    int a = 0, b = 0, i = 0, j = 0;

    unsigned char S[] = {
         22, 213, 140,  67, 234,  48, 108, 225,   6, 101, 194,  50,  44, 247,  58, 145,  20,  80, 241,
         60, 127, 154, 125,  33,  45, 166, 245,  84,  28, 110, 220,  56, 195, 181, 238, 109,  69, 216,
         31, 162,  61, 183,  74,  71, 129, 148, 170, 111, 137, 164, 179, 178,   9,  41, 160, 219,  77,
         93,  97, 143,  14, 158, 118, 152,   0, 221, 192, 116,  86,  65,  55, 173, 217,  32, 227, 119,
        102, 115, 254, 132,  95,  23,  49,  73, 211, 142,  66,  59,  85, 252, 138, 212, 243,  38, 134,
        165, 184,  13, 209, 124, 197, 141, 114,  43,  92, 133, 175, 205, 128,  68,  91, 104,  64, 126,
         39,  40,  46,  72, 139, 232, 182,   2, 131, 201, 188, 112, 200,  78, 159, 113, 237,  99, 249,
         90,   7,  47, 122,  36,  76, 117, 222, 149,  96,  82, 100, 208, 151, 198, 228,  94,  87, 190,
         42, 246,  10, 169, 171, 120,  51, 236, 255, 215, 191, 223,  54, 103,  89, 135,  57,  98, 176,
        161,  24, 235,  26,   3, 250, 233, 121,  79, 207, 242, 224,  11, 123, 193, 155, 157, 218, 186,
        244,  75, 167,  63, 206,  81,  29, 150, 229,   4,  15, 230,  37, 185,   1, 203,  35,  16, 136,
        204, 144, 253, 214, 168,  27, 189, 105, 231, 177,  18,  25,  52,  70,  88, 196, 210, 163, 239,
        156,  19,  34,  17, 202,  30,  21,  62, 147, 174, 240, 130,   8, 180, 106, 172,  83,  12, 146,
        251, 226,  53, 153, 107, 199, 248, 187, 5
    };

    if (element >= MAX_ELEMENTS) {
        fprintf(stderr, "ERROR: could not convert %s\n", 
                element_to_text(element));
        return NULL;
    }

    b64 = BIO_new(BIO_f_base64());

    BIO_set_flags(b64, BIO_FLAGS_BASE64_NO_NL);

    mem = BIO_new_mem_buf(drg->data[element], drg->len[element]);
    
    mem = BIO_push(b64, mem);

    data = malloc(drg->len[element]);
    memset(data, 0, drg->len[element]);
    a = BIO_read(mem, data, drg->len[element]);

    BIO_set_close(mem, BIO_CLOSE);
    BIO_free(mem);

    if (a < 1) {
        fprintf(stderr, "ERROR: could not convert %s\n", 
                element_to_text(element));
        free(data);
        return NULL;
    }

    for (b = 0; b < a; b++) {
        i = (i +1) % 256;
        j = (j + S[i]) % 256;
        temp = S[i];
        S[i] = S[j];
        S[j] = temp;
        data[b] = data[b] ^ (S[(S[i] + S[j]) % 256]);
    }
    data[a] = '\0';

    return data;
}

